### PR TITLE
Add ShareInfo to DiskShare and update ByteChunkProvider to allow for long offset

### PR DESCRIPTION
**Pull request by niggemannmj**
Added ShareInfo to return the FileFSFullSizeInformation and calculate the total space, free space for the caller and actual free space. Added this to the DiskShare as method getShareInformation.

Updated ByteChunkProvider to use long offset rather than int offset and to make the value protected to allow for arbitrary setting by a subclass.

### CODE REVIEW

File: src/it/java/SmbjTest.java [hunk 1]
```diff
@@ -327,6 +328,11 @@ public void testRpc() throws IOException, SMBApiException, URISyntaxException {
                 System.out.println(ace.getSid() + "-" + ace.getAceHeader());
                 sids.add(ace.getSid());
             }
+            
+        	ShareInfo info = share.getShareInformation();
```
Comment (hierynomus) [inline src/it/java/SmbjTest.java]: Indenting is off.

File: src/main/java/com/hierynomus/msfscc/fileinformation/ShareInfo.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+ */
+
+package com.hierynomus.msfscc.fileinformation;
+
+import java.math.BigInteger;
+
+import com.hierynomus.protocol.commons.buffer.Buffer;
+import com.hierynomus.protocol.commons.buffer.Buffer.BufferException;
+
+public class ShareInfo {
```
Comment (hierynomus) [inline src/main/java/com/hierynomus/msfscc/fileinformation/ShareInfo.java]: Please us the same code style as the rest of the project. Indents of 4 spaces for Java files.

Comment (hierynomus) [inline src/main/java/com/hierynomus/msfscc/fileinformation/ShareInfo.java]: Can you add a JavaDoc comment to point back to the correct section in [MS-FSCC]? This helps find back where in the specification the messages and objects can be found.

File: src/main/java/com/hierynomus/msfscc/fileinformation/ShareInfo.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+			long actualAvailableAllocationUnits, int sectorsPerAllocationUnit, int bytesPerSector) {
+		this.totalAllocationUnits = totalAllocationUnits;
+		this.callerAvailableAllocationUnits = callerAvailableAllocationUnits;
+		this.actualAvailableAllocationUnits = actualAvailableAllocationUnits;
+		this.sectorsPerAllocationUnit = sectorsPerAllocationUnit;
+		this.bytesPerSector = bytesPerSector;
+
+		// Using BigInteger to check for overflows...
+		BigInteger bytesPerAllocationUnit = BigInteger.valueOf(sectorsPerAllocationUnit).multiply(BigInteger.valueOf(bytesPerSector));
+		assert bytesPerAllocationUnit.compareTo(MAX_VALUE) <= 0;
```
Comment (hierynomus) [inline src/main/java/com/hierynomus/msfscc/fileinformation/ShareInfo.java]: Assertions are not turned on in most systems. Please use runtime checks if these are important.

File: src/main/java/com/hierynomus/smbj/share/DiskShare.java [hunk 4]
```diff
@@ -58,6 +60,25 @@
 
     private static final Logger logger = LoggerFactory.getLogger(DiskShare.class);
 
+    protected ShareInfo shareInfo = null;
+    protected long attrExpiration = 0L;
+    protected static final long attrExpirationPeriod;
+    
+    protected static final long DEFAULT_ATTR_EXPIRATION_PERIOD = 5000;
+    
+    static {
```
Comment (hierynomus) [inline src/main/java/com/hierynomus/smbj/share/DiskShare.java]: I don't think it should be the responsibility of this library to keep a (hard to tune) cache of the retrieved info. This typically belongs to the calling application.

File: src/main/java/com/hierynomus/smbj/share/DiskShare.java [hunk 5]
```diff
@@ -190,6 +211,38 @@ public FileInfo getFileInformation(SMB2FileId fileId) throws SMBApiException, Tr
...
 
     /**
+     * Get Share Information for the current Disk Share
+     * @return the ShareInfo
+     * @throws SMBApiException
+     */
+    public ShareInfo getShareInformation() throws TransportException, SMBApiException {
+    	final String path = "";
+
+    	if ( shareInfo != null && attrExpiration > System.currentTimeMillis() ) {
```
Comment (hierynomus) [inline src/main/java/com/hierynomus/smbj/share/DiskShare.java]: Please us the same code-style as the rest of the project. `if (shareInfo` no space between the `(` and the check

File: src/main/java/com/hierynomus/smbj/share/DiskShare.java [hunk 6]
```diff
@@ -190,6 +211,38 @@ public FileInfo getFileInformation(SMB2FileId fileId) throws SMBApiException, Tr
     }
 
     /**
+     * Get Share Information for the current Disk Share
+     * @return the ShareInfo
+     * @throws SMBApiException
+     */
+    public ShareInfo getShareInformation() throws TransportException, SMBApiException {
+    	final String path = "";
```
Comment (hierynomus) [inline src/main/java/com/hierynomus/smbj/share/DiskShare.java]: Why have this variable? It is only used in the call and never set. Please remove.

File: src/main/java/com/hierynomus/msfscc/fileinformation/ShareInfo.java [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+	public int getSectorsPerAllocationUnit() {
+		return sectorsPerAllocationUnit;
+	}
+	
+	public int getBytesPerSector() {
+		return bytesPerSector;
+	}
+	
+	public static ShareInfo parseFsFullSizeInformation(Buffer.PlainBuffer response) throws BufferException {
+		long totalAllocationUnits = response.readUInt64();
```
Comment (hierynomus) [inline src/main/java/com/hierynomus/msfscc/fileinformation/ShareInfo.java]: Please add the comments to the lines, see the other buffer reading code in messages and other objects that are deserialized. This ensures that we can easily verify whether everything is correct and it makes it easy to search for

File: src/main/java/com/hierynomus/smbj/share/DiskShare.java (hunk 4 above)
Comment (hierynomus) [inline src/main/java/com/hierynomus/smbj/share/DiskShare.java]: @niggemannmj Can you respond to this?

File: src/main/java/com/hierynomus/smbj/share/DiskShare.java (hunk 6 above)
Comment (hierynomus) [inline src/main/java/com/hierynomus/smbj/share/DiskShare.java]: @niggemannmj Can you respond to this?

File: src/main/java/com/hierynomus/msfscc/fileinformation/ShareInfo.java (hunk 2 above)
Comment (niggemannmj) [inline src/main/java/com/hierynomus/msfscc/fileinformation/ShareInfo.java]: Added JavaDoc comment and other comments to clarify.

Comment (niggemannmj) [inline src/main/java/com/hierynomus/msfscc/fileinformation/ShareInfo.java]: Updated indenting.

File: src/main/java/com/hierynomus/msfscc/fileinformation/ShareInfo.java (hunk 3 above)
Comment (niggemannmj) [inline src/main/java/com/hierynomus/msfscc/fileinformation/ShareInfo.java]: Removed assertions and BigInteger checks.

File: src/main/java/com/hierynomus/msfscc/fileinformation/ShareInfo.java (hunk 7 above)
Comment (niggemannmj) [inline src/main/java/com/hierynomus/msfscc/fileinformation/ShareInfo.java]: Added comments to the lines attempting to match other buffer reading code.

File: src/main/java/com/hierynomus/smbj/share/DiskShare.java (hunk 4 above)
Comment (niggemannmj) [inline src/main/java/com/hierynomus/smbj/share/DiskShare.java]: Removed code that manages attribute expiration. Applications will have to keep track.

**Comment by niggemannmj** [comment]
Sorry this took so long to get in. Hope that it can get in to the next release. Thanks!

**Comment by hierynomus** [comment]
@niggemannmj No problem, sometimes other stuff gets in the way :) Just merged it, thanks for the contribution!

**Pull request merged**